### ISSUE TITLE
[BUG] Invalid path when logging image to GCS on step 23

### ISSUE THREAD

**Issue by dails08** [mlflow/mlflow#15718]
### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Other

### MLflow version

- Client: 2.22.0
- Tracking server: 2.22.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 25.04
- **Python version**: 3.13
- **yarn version, if running the dev UI**: NA

### Describe the problem

When running mlflow.log_image(image_data, key=key, step=step), on step 23, mlflow consistently throws an INVALID PATH error. Uncommenting the code that skips specifically step 23 results in no error.

Artifacts are tracked in GCS, so it may be a permissions error, but no other steps error and the images (png and webp) log and store fine. Backend is GCP CloudSQL postgres, so it could be a failure there, but again, no other steps error out. All other run information is persisted.

This one error also seems to invalidate some logic checking for the mlflow UI, since the images (all of which except for step 23 are persisted and viewable individually) no longer show up as a slider-enabled series of plots in the model metrics tab, nor do they populate in the image grid option Images dropdown.

### Tracking information

```shell
System information: Linux #19~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Feb 17 11:51:52 UTC 2
Python version: 3.12.3
MLflow version: 2.22.0
MLflow module location: /home/chris/Desktop/repos/redacted/research/mlflow/env-mlflow/lib/python3.12/site-packages/mlflow/__init__.py
Tracking URI: http://xx.xx.xxx.xxx:8885/
Registry URI: http://xx.xx.xxx.xxx:8885/
MLflow environment variables: 
  MLFLOW_TRACKING_URI: http://xx.xx.xxx.xxx:8885/
MLflow dependencies: 
  Flask: 3.1.0
  Jinja2: 3.1.6
  alembic: 1.15.2
  docker: 7.1.0
  fastapi: 0.115.12
  google-cloud-storage: 3.1.0
  graphene: 3.4.3
  gunicorn: 23.0.0
  markdown: 3.8
  matplotlib: 3.10.1
  mlflow-skinny: 2.22.0
  numpy: 2.2.5
  pandas: 2.2.3
  pyarrow: 19.0.1
  scikit-learn: 1.6.1
  scipy: 1.15.2
  sqlalchemy: 2.0.40
  uvicorn: 0.34.2
```

### Code to reproduce issue

```
import torch
from torch import nn
from torch.utils.data import Dataset
from torchvision.transforms import ToTensor
import matplotlib.pyplot as plt

from google.cloud import storage
import numpy as np
from PIL import Image

from loguru import logger

import mlflow

import os
from io import BytesIO

mlflow.set_tracking_uri("http://xx.xx.xxx.xxx:8885")

mlflow.search_experiments() # to verify connection and credentials

rng = np.random.default_rng()

def plot_estimates_at_step(model, step):
    fig, ax = plt.subplots(figsize = (5,5))
    preds_X = np.linspace(0,100,10).reshape((-1, 1))
    preds_dataset = model(preds_X).detach().numpy()
    ax.scatter(raw_data.X, raw_data.Y)
    ax.scatter(preds_X, preds_dataset)
    ax.set_title(f"Step {step}")
    plt.tight_layout()
    buf = BytesIO()
    plt.savefig(buf)
    plt.close()

    buf.seek(0)
    img = Image.open(buf)
    return img

class SimpleLinear(nn.Module):
    def __init__(self, input_shape = 5, output_shape = 1):
        super(SimpleLinear, self).__init__()


        self.model = nn.Sequential(
            nn.Linear(input_shape, 6, dtype = torch.float32),
            nn.ReLU(),
            nn.Linear(6, output_shape, dtype = torch.float32)
        )
        # self.model = nn.Linear(input_shape, output_shape)

    def forward(self, x):
        x = torch.Tensor(x)
        output = self.model(x)
        # print(output)
        return output


class SimpleDataset(Dataset):
    def __init__(self):
        self.X = torch.from_numpy(np.linspace(0,100,200, dtype = np.float32))
        eps = rng.normal(size = 200)
        self.Y = 1.12 * self.X**2 + 7.75 + eps
        self.Y = torch.from_numpy(np.asarray(self.Y, dtype = np.float32))

    def __len__(self):
        return self.X.shape[0]

    def __getitem__(self, idx):
        return self.X[idx], self.Y[idx]

raw_data = SimpleDataset()

linear_model = SimpleLinear(input_shape=1, output_shape=1)
optimizer = torch.optim.Adam(linear_model.parameters(), lr = 0.01)
loss_fn = nn.MSELoss()

loader = torch.utils.data.DataLoader( dataset = raw_data, batch_size = 9, shuffle = True)

num_epochs = 4
with mlflow.start_run(
    run_name="Step 23 bug MWE",
    description="Demonstrate consistent error when logging an image at step 23",
) as outer_run:
    logger.info("Starting run")
    mlflow.log_param("model", "simple_linear")
    mlflow.log_param("dataset", "SimpleDataset")
    mlflow.log_param("epochs", "5")
    mlflow.log_param("optimizer", "adam")
    mlflow.log_param("lr", 0.01)
    mlflow.log_param("loss", "mse")
    step = 0
    for epoch in range(num_epochs):
        logger.info(f"Starting epoch {epoch}")
        for ix, (x_batch, y_batch) in enumerate(loader):
            logger.debug(f"Step {step}")
            x_batch = x_batch.reshape((-1,1))
            y_batch = y_batch.reshape((-1, 1))
            preds = linear_model(x_batch)
            loss = loss_fn(preds, y_batch)
    
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            mlflow.log_metric("loss", np.mean(loss.detach().numpy()), step = step)
            # if step != 23:
            #     mlflow.log_image( plot_estimates_at_step(linear_model, step), key="estimate", step = step)
            mlflow.log_image( plot_estimates_at_step(linear_model, step), key="estimate", step = step)
            step += 1


```

### Stack trace

```
2025/05/13 19:03:04 ERROR mlflow.utils.async_logging.async_artifacts_logging_queue: Failed to log artifact estimates%step%23%timestamp%1747177383576%kf4f30c9-adf8-4c92-8e06-687589887029%compressed.webp. Exception: 400 Client Error: BAD REQUEST for url: http://xx.xx.xxx.xxx:8885/api/2.0/mlflow-artifacts/artifacts/445020544189088996/3c66290bf7fd47bc97922f349a0b193f/artifacts/images/estimates%25step%2523%25timestamp%251747177383576%25kf4f30c9-adf8-4c92-8e06-687589887029%25compressed.webp. Response text: {"error_code": "INVALID_PARAMETER_VALUE", "message": "Invalid path"}
2025/05/13 19:03:05 ERROR mlflow.utils.async_logging.async_artifacts_logging_queue: Failed to log artifact estimates%step%23%timestamp%1747177383576%kf4f30c9-adf8-4c92-8e06-687589887029.png. Exception: API request to http://xx.xx.xxx.xxx:8885/api/2.0/mlflow-artifacts/artifacts/445020544189088996/3c66290bf7fd47bc97922f349a0b193f/artifacts/images/estimates%step%23%timestamp%1747177383576%kf4f30c9-adf8-4c92-8e06-687589887029.png failed with exception ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

### Other info / logs

```
(env-mlflow) redacted@mlflow-server:~$ mlflow server --host 0.0.0.0 --port 8885 --artifacts-destination gs://redacted/mlflow --app-name basic-auth
[2025-05-13 17:29:12 +0000] [1753] [INFO] Starting gunicorn 23.0.0
[2025-05-13 17:29:12 +0000] [1753] [INFO] Listening at: http://0.0.0.0:8885 (1753)
[2025-05-13 17:29:12 +0000] [1753] [INFO] Using worker: sync
[2025-05-13 17:29:12 +0000] [1754] [INFO] Booting worker with pid: 1754
[2025-05-13 17:29:12 +0000] [1755] [INFO] Booting worker with pid: 1755
[2025-05-13 17:29:12 +0000] [1756] [INFO] Booting worker with pid: 1756
[2025-05-13 17:29:12 +0000] [1757] [INFO] Booting worker with pid: 1757
2025/05/13 17:29:17 WARNING mlflow.server.auth: This feature is still experimental and may change in a future release without warning
2025/05/13 17:29:17 WARNING mlflow.server.auth: This feature is still experimental and may change in a future release without warning
2025/05/13 17:29:17 WARNING mlflow.server.auth: This feature is still experimental and may change in a future release without warning
2025/05/13 17:29:17 WARNING mlflow.server.auth: This feature is still experimental and may change in a future release without warning
```

### What component(s) does this bug affect?

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by dails08** [comment]
This feels related to https://github.com/mlflow/mlflow/issues/12151, as images, even when skipping step 23, fail to register both in the slider-enabled model metric and in the dropdown for creating an image grid.

**Comment by TomeHirata** [comment]
Thank you for the report. %23 means "#" in percent encoding, so I guess the rest of the string is recognized as a URL fragment. In a few months, we will try to replace the separator with something else to fix the log_image behavior completely.

**Comment by ceranco** [comment]
Hi @TomeHirata, has there been any update on this?
